contracts/src/registry/EnhancedAccessControl.sol

```diff
@@ -0,0 +1,302 @@
+     */
+    function hasRootRoles(uint256 rolesBitmap, address account) public view virtual returns (bool) {
+        return (_roles[ROOT_RESOURCE][account] & rolesBitmap) != 0;
+    }
+
+    /**
+     * @dev Returns `true` if `account` has been granted all roles in `resource`.
+     */
+    function hasRoles(bytes32 resource, uint256 rolesBitmap, address account) public view virtual returns (bool) {
+        return ((_roles[resource][account] & rolesBitmap) != 0) || hasRootRoles(rolesBitmap, account);
+    }
+
+
+    /**
+     * @dev Returns the admin role that controls `role`. See {grantRole} and
+     * {revokeRole}.
+     *
+     * To change a role's admin, use {_setRoleAdmin}.
+     */
+    function getRoleAdmin(uint256 role) public view virtual returns (uint256) {
+        return _adminRoles[role];
+    }
+
+    /**
+     * @dev Returns the locked roles bitmap for a resource.
+     */
+    function getLockedRoles(bytes32 resource) public view virtual returns (uint256) {
+        return _lockedRoles[resource];
+    }
+
+    /**
+     * @dev Grants role to `account`.
+     *
+     * If `account` had not been already granted `role`, emits a {RoleGranted}
+     * event.
+     *
+     * Requirements:
+     *
+     * - the caller must have ``role``'s admin role.
+     *
+     * May emit a {RoleGranted} event.
+     */
+    function grantRole(bytes32 resource, uint256 role, address account) public virtual canGrantRole(resource, role) returns (bool) {
+        return _grantRoles(resource, role, account);
+    }
+
+
+    /**
+     * @dev Revokes `role` from `account`.
+     *
+     * If `account` had been granted `role`, emits a {RoleRevoked} event.
+     *
+     * Requirements:
+     *
+     * - the caller must have ``role``'s admin role.
+     *
+     * May emit a {RoleRevoked} event.
+     */
+    function revokeRole(bytes32 resource, uint256 role, address account) public virtual canGrantRole(resource, role) returns (bool) {
+        return _revokeRoles(resource, role, account);
+    }
+
+
+    /**
+     * @dev Revokes `role` from the calling account.
+     *
+     * Roles are often managed via {grantRole} and {revokeRole}: this function's
+     * purpose is to provide a mechanism for accounts to lose their privileges
+     * if they are compromised (such as when a trusted device is misplaced).
+     *
+     * If the calling account had been revoked `role`, emits a {RoleRevoked}
+     * event.
+     *
+     * Requirements:
+     *
+     * - the caller must be `callerConfirmation`.
+     *
+     * May emit a {RoleRevoked} event.
+        *
+     * Returns `true` if the role was revoked, `false` otherwise.
+     */
+    function renounceRole(bytes32 resource, uint256 role, address callerConfirmation) public virtual returns (bool) {
+        if (callerConfirmation != _msgSender()) {
+            revert EnhancedAccessControlBadConfirmation();
+        }
+
+        return _revokeRoles(resource, role, callerConfirmation);
+    }
```

Comment:
I think we should omit this; if an account is an admin for the role, they can revoke both the role and their own admin status, but having this makes it impossible to 'lock open' a permission.

Reply:
Revoke all admin access over a role, without revoking the role itself for everyone - meaning it can still be done, but the set of accounts that can do it can't be changed any longer.